File: core/src/main/java/com/pmcl/core/install/IntegrityChecker.java
package com.pmcl.core.install;

import com.pmcl.core.LauncherConfig;
import com.google.gson.JsonObject;
import com.google.gson.JsonParser;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.security.MessageDigest;
import java.util.ArrayList;
import java.util.List;

/**
 * 启动前完整性校验：根据版本 JSON 校验 client.jar 与所有 libraries 的 SHA1。
 * <p>
 * 缺失或哈希不匹配的文件会被收集到 {@link Result} 中，UI 可提示用户重新下载。
 */
public final class IntegrityChecker {

    private final LauncherConfig config;

    public IntegrityChecker(LauncherConfig config) {
        this.config = config;
    }

    public static final class Result {
        private final List<String> missing = new ArrayList<>();
        private final List<String> hashMismatch = new ArrayList<>();
        private final List<String> ok = new ArrayList<>();
        /** 存在但无法哈希校验（无 sha1 / maven 仅路径） */
        private final List<String> unverifiable = new ArrayList<>();

        public List<String> getMissing() { return missing; }
        public List<String> getHashMismatch() { return hashMismatch; }
        public List<String> getOk() { return ok; }
        public List<String> getUnverifiable() { return unverifiable; }

        public boolean isOk() { return missing.isEmpty() && hashMismatch.isEmpty(); }

        public int getIssueCount() { return missing.size() + hashMismatch.size(); }
    }

    /**
     * 校验指定版本。
     */
    public Result check(String versionId) throws IOException {
        Result result = new Result();
        Path versionDir = config.getVersionsDir().resolve(versionId);
        Path versionJson = versionDir.resolve(versionId + ".json");
        if (!Files.exists(versionJson)) {
            result.getMissing().add("versions/" + versionId + "/" + versionId + ".json");
            return result;
        }

        JsonObject root = JsonParser.parseString(Files.readString(versionJson, java.nio.charset.StandardCharsets.UTF_8)).getAsJsonObject();

        // client.jar
        if (root.has("downloads")) {
            JsonObject dl = root.getAsJsonObject("downloads");
            if (dl.has("client")) {
                JsonObject client = dl.getAsJsonObject("client");
                Path clientJar = versionDir.resolve(versionId + ".jar");
                if (client.has("sha1") && !client.get("sha1").isJsonNull()) {
                    verifyFile(clientJar, client.get("sha1").getAsString(), result);
                } else if (Files.exists(clientJar)) {
                    result.getUnverifiable().add(clientJar + " (client.jar 无 sha1)");
                } else {
                    result.getMissing().add(clientJar.toString());
                }
            }
        }

        // libraries
        if (root.has("libraries")) {
            for (var e : root.getAsJsonArray("libraries")) {
                JsonObject lib = e.getAsJsonObject();
                if (!lib.has("downloads")) {
                    // M84: Fabric/Forge/NeoForge 第三方库格式（只有顶层 name + url，无 downloads/sha1）
                    // 按 maven 规则拼接路径，仅校验文件存在（无 SHA1 可比对）→ unverifiable
                    if (lib.has("name") && !lib.get("name").isJsonNull()) {
                        Library parsed = Library.parse(lib);
                        String path = parsed.getPath();
                        if (!path.isEmpty()) {
                            Path libFile = config.getLibrariesDir().resolve(path);
                            if (!Files.exists(libFile)) {
                                result.getMissing().add(libFile.toString());
                            } else {
                                result.getUnverifiable().add(libFile + " (无 sha1)");
                            }
                        }
                    }
                    continue;
                }
                JsonObject downloads = lib.getAsJsonObject("downloads");
                if (downloads.has("artifact")) {
                    JsonObject art = downloads.getAsJsonObject("artifact");
                    if (art.has("path") && !art.get("path").isJsonNull()) {
                        Path libFile = config.getLibrariesDir().resolve(art.get("path").getAsString());
                        if (art.has("sha1") && !art.get("sha1").isJsonNull()) {
                            verifyFile(libFile, art.get("sha1").getAsString(), result);
                        } else if (Files.exists(libFile)) {
                            result.getUnverifiable().add(libFile + " (artifact 无 sha1)");
                        } else {
                            result.getMissing().add(libFile.toString());
                        }
                    }
                }
                // native classifiers
                if (downloads.has("classifiers")) {
                    JsonObject cl = downloads.getAsJsonObject("classifiers");
                    for (var ce : cl.entrySet()) {
                        JsonObject a = ce.getValue().getAsJsonObject();
                        if (!a.has("path") || a.get("path").isJsonNull()) continue;
                        Path libFile = config.getLibrariesDir().resolve(a.get("path").getAsString());
                        if (a.has("sha1") && !a.get("sha1").isJsonNull()) {
                            verifyFile(libFile, a.get("sha1").getAsString(), result);
                        } else if (Files.exists(libFile)) {
                            result.getUnverifiable().add(libFile + " (classifier 无 sha1)");
                        } else {
                            result.getMissing().add(libFile.toString());
                        }
                    }
                }
            }
        }

        // assets：校验 assetIndex 文件 + objects 哈希（hash 即 sha1）
        checkAssets(root, result);

        return result;
    }

    private void checkAssets(JsonObject root, Result result) {
        if (!root.has("assetIndex") || root.get("assetIndex").isJsonNull()) return;
        JsonObject ai = root.getAsJsonObject("assetIndex");
        String id = ai.has("id") && !ai.get("id").isJsonNull() ? ai.get("id").getAsString() : null;
        if (id == null || id.isBlank()) return;
        Path indexPath = config.getAssetsDir().resolve("indexes").resolve(id + ".json");
        if (ai.has("sha1") && !ai.get("sha1").isJsonNull()) {
            verifyFile(indexPath, ai.get("sha1").getAsString(), result);
        } else if (Files.exists(indexPath)) {
            result.getUnverifiable().add(indexPath + " (assetIndex 无 sha1)");
        } else {
            result.getMissing().add(indexPath.toString());
            return;
        }
        if (!Files.exists(indexPath)) return;
        try {
            JsonObject idx = JsonParser.parseString(
                    Files.readString(indexPath, java.nio.charset.StandardCharsets.UTF_8)).getAsJsonObject();
            if (!idx.has("objects")) return;
            JsonObject objects = idx.getAsJsonObject("objects");
            Path objectsDir = config.getAssetsDir().resolve("objects");
            for (var e : objects.entrySet()) {
                JsonObject obj = e.getValue().getAsJsonObject();
                if (!obj.has("hash") || obj.get("hash").isJsonNull()) {
                    result.getUnverifiable().add("assets object " + e.getKey() + " (无 hash)");
                    continue;
                }
                String hash = obj.get("hash").getAsString();
                if (hash.length() < 2) {
                    result.getUnverifiable().add("assets object " + e.getKey() + " (非法 hash)");
                    continue;
                }
                Path file = objectsDir.resolve(hash.substring(0, 2)).resolve(hash);
                verifyFile(file, hash, result);
            }
        } catch (Exception ex) {
            result.getHashMismatch().add(indexPath + " (解析失败: " + ex.getMessage() + ")");
        }
    }

    private void verifyFile(Path file, String expectedSha1, Result result) {
        if (!Files.exists(file)) {
            result.getMissing().add(file.toString());
            return;
        }
        try {
            String actual = sha1(file);
            if (!actual.equalsIgnoreCase(expectedSha1)) {
                result.getHashMismatch().add(file.toString() + " (期望=" + expectedSha1 + " 实际=" + actual + ")");
            } else {
                result.getOk().add(file.toString());
            }
        } catch (IOException e) {
            result.getHashMismatch().add(file.toString() + " (计算哈希失败: " + e.getMessage() + ")");
        }
    }

    private static String sha1(Path file) throws IOException {
        try {
            MessageDigest md = MessageDigest.getInstance("SHA-1");
            try (var is = Files.newInputStream(file)) {
                byte[] buf = new byte[8192];
                int n;
                while ((n = is.read(buf)) != -1) md.update(buf, 0, n);
            }
            byte[] digest = md.digest();
            StringBuilder sb = new StringBuilder();
            // H13: b & 0xff 防止 byte 符号扩展为 int 时产生 ffffffff 而非 ff
            for (byte b : digest) sb.append(String.format("%02x", b & 0xff));
            return sb.toString();
        } catch (Exception e) {
            throw new IOException("SHA1 计算失败", e);
        }
    }
}

File: pcmlAndroid/core/src/main/java/com/lash/pmcl/core/install/IntegrityChecker.kt
package com.lash.pmcl.core.install

import com.google.gson.JsonObject
import com.google.gson.JsonParser
import com.lash.pmcl.core.paths.PmclPaths
import com.lash.pmcl.core.util.FileUtils
import java.io.IOException
import java.nio.charset.StandardCharsets
import java.nio.file.Files
import java.nio.file.Path
import java.security.MessageDigest

/**
 * 启动前完整性校验：根据版本 JSON 校验 client.jar 与所有 libraries 的 SHA1。
 * <p>
 * 缺失或哈希不匹配的文件会被收集到 [Result] 中，UI 可提示用户重新下载。
 *
 * Android 版本：通过 [PmclPaths] 提供路径，无 LauncherConfig 依赖。
 */
class IntegrityChecker(private val paths: PmclPaths) {

    class Result {
        val missing: MutableList<String> = ArrayList()
        val hashMismatch: MutableList<String> = ArrayList()
        val ok: MutableList<String> = ArrayList()
        /** 存在但无法哈希校验（无 sha1 / maven 仅路径） */
        val unverifiable: MutableList<String> = ArrayList()

        val isOk: Boolean get() = missing.isEmpty() && hashMismatch.isEmpty()
        val issueCount: Int get() = missing.size + hashMismatch.size
    }

    /**
     * 校验指定版本。
     */
    @Throws(IOException::class)
    fun check(versionId: String): Result {
        val result = Result()
        val versionDir = paths.versions.resolve(versionId)
        val versionJson = versionDir.resolve("$versionId.json")
        if (!Files.exists(versionJson)) {
            result.missing.add("versions/$versionId/$versionId.json")
            return result
        }

        val root = JsonParser.parseString(
            FileUtils.readString(versionJson, StandardCharsets.UTF_8)
        ).asJsonObject

        // client.jar
        if (root.has("downloads")) {
            val dl = root.getAsJsonObject("downloads")
            if (dl.has("client")) {
                val client = dl.getAsJsonObject("client")
                val clientJar = versionDir.resolve("$versionId.jar")
                if (client.has("sha1") && !client.get("sha1").isJsonNull) {
                    verifyFile(clientJar, client.get("sha1").asString, result)
                } else if (Files.exists(clientJar)) {
                    result.unverifiable.add("$clientJar (client.jar 无 sha1)")
                } else {
                    result.missing.add(clientJar.toString())
                }
            }
        }

        // libraries
        if (root.has("libraries")) {
            for (e in root.getAsJsonArray("libraries")) {
                val lib = e.asJsonObject
                if (!lib.has("downloads")) {
                    // Fabric/Forge/NeoForge 第三方库格式（只有顶层 name + url，无 downloads/sha1）
                    if (lib.has("name") && !lib.get("name").isJsonNull) {
                        val parsed = Library.parse(lib)
                        val path = parsed.getPath()
                        if (path.isNotEmpty()) {
                            val libFile = paths.libraries.resolve(path)
                            if (!Files.exists(libFile)) {
                                result.missing.add(libFile.toString())
                            } else {
                                result.unverifiable.add("$libFile (无 sha1)")
                            }
                        }
                    }
                    continue
                }
                val downloads = lib.getAsJsonObject("downloads")
                if (downloads.has("artifact")) {
                    val art = downloads.getAsJsonObject("artifact")
                    if (art.has("path") && !art.get("path").isJsonNull) {
                        val libFile = paths.libraries.resolve(art.get("path").asString)
                        if (art.has("sha1") && !art.get("sha1").isJsonNull) {
                            verifyFile(libFile, art.get("sha1").asString, result)
                        } else if (Files.exists(libFile)) {
                            result.unverifiable.add("$libFile (artifact 无 sha1)")
                        } else {
                            result.missing.add(libFile.toString())
                        }
                    }
                }
                // native classifiers
                if (downloads.has("classifiers")) {
                    val cl = downloads.getAsJsonObject("classifiers")
                    for ((_, ceVal) in cl.entrySet()) {
                        val a = ceVal.asJsonObject
                        if (!a.has("path") || a.get("path").isJsonNull) continue
                        val libFile = paths.libraries.resolve(a.get("path").asString)
                        if (a.has("sha1") && !a.get("sha1").isJsonNull) {
                            verifyFile(libFile, a.get("sha1").asString, result)
                        } else if (Files.exists(libFile)) {
                            result.unverifiable.add("$libFile (classifier 无 sha1)")
                        } else {
                            result.missing.add(libFile.toString())
                        }
                    }
                }
            }
        }

        // assets
        checkAssets(root, result)

        return result
    }

    private fun checkAssets(root: JsonObject, result: Result) {
        if (!root.has("assetIndex") || root.get("assetIndex").isJsonNull) return
        val ai = root.getAsJsonObject("assetIndex")
        val id = if (ai.has("id") && !ai.get("id").isJsonNull) ai.get("id").asString else null
        if (id.isNullOrBlank()) return
        val indexPath = paths.assets.resolve("indexes").resolve("$id.json")
        if (ai.has("sha1") && !ai.get("sha1").isJsonNull) {
            verifyFile(indexPath, ai.get("sha1").asString, result)
        } else if (Files.exists(indexPath)) {
            result.unverifiable.add("$indexPath (assetIndex 无 sha1)")
        } else {
            result.missing.add(indexPath.toString())
            return
        }
        if (!Files.exists(indexPath)) return
        try {
            val idx = JsonParser.parseString(
                FileUtils.readString(indexPath, StandardCharsets.UTF_8)
            ).asJsonObject
            if (!idx.has("objects")) return
            val objects = idx.getAsJsonObject("objects")
            val objectsDir = paths.assets.resolve("objects")
            for ((key, eVal) in objects.entrySet()) {
                val obj = eVal.asJsonObject
                if (!obj.has("hash") || obj.get("hash").isJsonNull) {
                    result.unverifiable.add("assets object $key (无 hash)")
                    continue
                }
                val hash = obj.get("hash").asString
                if (!hash.matches(Regex("[0-9a-fA-F]{40}"))) {
                    result.unverifiable.add("assets object $key (非法 hash)")
                    continue
                }
                val file = objectsDir.resolve(hash.substring(0, 2)).resolve(hash).normalize()
                if (!file.startsWith(objectsDir.toAbsolutePath().normalize())) {
                    result.hashMismatch.add("assets object $key (路径越界)")
                    continue
                }
                verifyFile(file, hash, result)
            }
        } catch (ex: Exception) {
            result.hashMismatch.add("$indexPath (解析失败: ${ex.message})")
        }
    }

    private fun verifyFile(file: Path, expectedSha1: String, result: Result) {
        if (!Files.exists(file)) {
            result.missing.add(file.toString())
            return
        }
        try {
            val actual = sha1(file)
            if (!actual.equals(expectedSha1, ignoreCase = true)) {
                result.hashMismatch.add("$file (期望=$expectedSha1 实际=$actual)")
            } else {
                result.ok.add(file.toString())
            }
        } catch (e: IOException) {
            result.hashMismatch.add("$file (计算哈希失败: ${e.message})")
        }
    }

    companion object {
        @Throws(IOException::class)
        private fun sha1(file: Path): String {
            return try {
                val md = MessageDigest.getInstance("SHA-1")
                Files.newInputStream(file).use { inp ->
                    val buf = ByteArray(8192)
                    while (true) {
                        val n = inp.read(buf)
                        if (n == -1) break
                        md.update(buf, 0, n)
                    }
                }
                val digest = md.digest()
                val sb = StringBuilder(digest.size * 2)
                // H13: b and 0xff 防止 byte 符号扩展为 int 时产生 ffffffff 而非 ff
                for (b in digest) sb.append(String.format("%02x", b.toInt() and 0xff))
                sb.toString()
            } catch (e: Exception) {
                throw IOException("SHA1 计算失败", e)
            }
        }
    }
}
